### ISSUE TITLE
BaseSpline.cpp:234 Assertion `dx[ix >0' failed.

### ISSUE THREAD

**Issue by kaipot** [percolator/percolator#177]
This was reported as a Crux issue but it occurs with standalone Percolator also. When compiled with debug mode, Percolator halts at the assert; in release mode it completes but the outputs have nan scores:

./percolator --results-peptides crux-output/percolator.target.peptides.txt --decoy-results-peptides crux-output/percolator.decoy.peptides.txt --results-psms crux-output/percolator.target.psms.txt --decoy-results-psms crux-output/percolator.decoy.psms.txt --verbose
 2 --protein-decoy-pattern decoy_ --seed 1 --subset-max-train 0 --trainFDR 0.01 --testFDR 0.01 --maxiter 10 --search-input auto --no-schema-validation --protein-enzyme trypsin --post-processing-tdc ~/documents/crux3/src/make-pin.pin
Percolator version 3.01.nightly-14-298ec65, Build Date Feb 16 2017 15:21:19
Copyright (c) 2006-9 University of Washington. All rights reserved.
Written by Lukas Käll (lukall@u.washington.edu) in the
Department of Genome Sciences at the University of Washington.
Issued command:
./percolator --results-peptides crux-output/percolator.target.peptides.txt --decoy-results-peptides crux-output/percolator.decoy.peptides.txt --results-psms crux-output/percolator.target.psms.txt --decoy-results-psms crux-output/percolator.decoy.psms.txt --verbose 2 --protein-decoy-p
attern decoy_ --seed 1 --subset-max-train 0 --trainFDR 0.01 --testFDR 0.01 --maxiter 10 --search-input auto --no-schema-validation --protein-enzyme trypsin --post-processing-tdc /home/kaipo/documents/crux3/src/make-pin.pin
Started Thu Feb 16 15:21:51 2017
Hyperparameters: selectionFdr=0.01, Cpos=0, Cneg=0, maxNiter=10
Reading tab-delimited input from datafile /home/kaipo/documents/crux3/src/make-pin.pin
Features:
lnrSp deltLCn deltCn XCorr Sp IonFrac PepLen Charge1 Charge2 Charge3 Charge4 Charge5 enzN enzC enzInt lnNumDSP dM absdM
Found 4983556 PSMs
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Reached strange feature with val=inf at col=4
Separate target and decoy search inputs detected, using target-decoy competition on Percolator scores.
Train/test set contains 2491778 positives and 2491778 negatives, size ratio=1 and pi0=1
Selecting Cpos by cross-validation.
Selecting Cneg by cross-validation.
Split 1:        Selected feature 4 as initial search direction. Could separate 907 training set positives in that direction.
Split 2:        Selected feature 4 as initial search direction. Could separate 933 training set positives in that direction.
Split 3:        Selected feature 4 as initial search direction. Could separate 851 training set positives in that direction.
Found 0 test set positives with q<0.01 in initial direction
Reading in data and feature calculation took 717.15 cpu seconds or 1041 seconds wall clock time.
---Training with Cpos selected by cross validation, Cneg selected by cross validation, fdr=0.01
Iteration 1:    Estimated 0 PSMs with q<0.01
Iteration 2:    Estimated 0 PSMs with q<0.01
Iteration 3:    Estimated 0 PSMs with q<0.01
Iteration 4:    Estimated 0 PSMs with q<0.01
Iteration 5:    Estimated 0 PSMs with q<0.01
Iteration 6:    Estimated 0 PSMs with q<0.01
Iteration 7:    Estimated 0 PSMs with q<0.01
Iteration 8:    Estimated 0 PSMs with q<0.01
Iteration 9:    Estimated 0 PSMs with q<0.01
Iteration 10:   Estimated 0 PSMs with q<0.01
Learned normalized SVM weights for the 3 cross-validation splits:
 Split1  Split2  Split3 FeatureName
-nan    -nan    -nan    lnrSp
-nan    -nan    -nan    deltLCn
-nan    -nan    -nan    deltCn
-nan    -nan    -nan    XCorr
-nan    -nan    -nan    Sp
-nan    -nan    -nan    IonFrac
-nan    -nan    -nan    PepLen
-nan    -nan    -nan    Charge1
-nan    -nan    -nan    Charge2
-nan    -nan    -nan    Charge3
-nan    -nan    -nan    Charge4
-nan    -nan    -nan    Charge5
-nan    -nan    -nan    enzN
-nan    -nan    -nan    enzC
-nan    -nan    -nan    enzInt
-nan    -nan    -nan    lnNumDSP
-nan    -nan    -nan    dM
-nan    -nan    -nan    absdM
-nan    -nan    -nan    m0
Found 0 test set PSMs with q<0.01.
No target score better than best decoy
Resetting score vector, using default vector
Split 1:        Selected feature 4 as initial search direction. Could separate 907 training set positives in that direction.
Split 2:        Selected feature 4 as initial search direction. Could separate 933 training set positives in that direction.
Split 3:        Selected feature 4 as initial search direction. Could separate 851 training set positives in that direction.
Selected best-scoring PSM per scan+expMass (target-decoy competition): 249522 target PSMs and 249184 decoy PSMs.
percolator: /home/kaipo/documents/percolator/src/BaseSpline.cpp:234: void BaseSpline::initiateQR(): Assertion `dx[ix] > 0' failed.

The pin file used was from here: https://drive.google.com/open?id=0BzgeYtbHb5hJUGFYSnBwNExNcjg

**Comment by wsnoble** [comment]
Actually, the stderr output does seem pretty informative: there seem to be some inf values in column 4. I've verified that these are in the input file in column 10, which contains the Sp score. Since the user generated this file from Crux, maybe the problem lies there. Kaipo, can you double back to the user and ask them for the command that generated this pin file?  Here is one of the problematic lines:

target_0_74821_2_5	1	74821	2216.0476	2216.1079	0.00000000	0.00000000	0.00846957	1.06293106	inf	0.15789473	20	0	1	0	0	0	1	1	0	6.68835449	-0.03017524	0.03017524	K.SQEASQTLLDSVYSHLPDLL.-	ENST00000296861.2|ENSG00000146072.6|OTTHUMG00000014796.1|OTTHUMT00000040814.1|TNFRSF21-001|TNFRSF21|655

I do think that Percolator should halt with an error when it encounters infinite features values, rather than continuing on and failing later.  Also, the error message should be improved to give the name of the problematic feature, and the line number where the inf occurs.

**Comment by wsnoble** [comment]
FWIW, the "inf" lines appear to be interspersed through the file.  Here are the first ten line numbers where they occur:

$ awk -F "\t" '$10 == "inf" {print NR}' make-pin.pin | head
1513706
1964208
1964282
2028220
2036617
2057718
2057732
2078897
2078915
2115959
